tiledb/sm/query/readers/attribute_order_validator.h

```diff
@@ -0,0 +1,581 @@
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/array_schema/dimension.h"
+#include "tiledb/sm/fragment/fragment_metadata.h"
+#include "tiledb/sm/query/readers/result_tile.h"
+
+#include <vector>
+
+using namespace tiledb::common;
+
+class ArraySchema;
+
+namespace tiledb::sm {
+
+class AttributeOrderValidatorStatusException : public StatusException {
+ public:
+  explicit AttributeOrderValidatorStatusException(const std::string& message)
+      : StatusException("ReaderBase", message) {
+  }
+};
+
+/**
+ * Utilitary function that returns if a value is contained in a non empty
+ * domain.
+ *
+ * @tparam Index type
+ * @param v Value to check.
+ * @param domain Pointer to the domain values.
+ * @return Is the value in the given domain or not.
+ */
+template <typename IndexType>
+uint8_t in_domain(IndexType v, const IndexType* domain) {
+  return v >= domain[0] && v <= domain[1];
+};
+
+class AttributeOrderValidator {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+  AttributeOrderValidator() = delete;
+
+  /**
+   * Construct a new Order Validation Data object.
+   *
+   * @param attribute_name Name of the attribute to validate.
+   * @param num_frags Number of fragments.
+   */
+  AttributeOrderValidator(const std::string& attribute_name, uint64_t num_frags)
+      : attribute_name_(attribute_name)
+      , result_tiles_to_load_(num_frags)
+      , min_validated_(num_frags, false)
+      , max_validated_(num_frags, false)
+      , fragment_to_compare_against_(num_frags, {nullopt, nullopt})
+      , tile_to_compare_against_(num_frags) {
+  }
+
+  /* ********************************* */
+  /*                 API               */
+  /* ********************************* */
+
+  /** Returns 'true' if tiles need to be loaded. */
+  inline bool need_to_load_tiles() {
+    for (auto& rt_map : result_tiles_to_load_) {
+      if (!rt_map.empty()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Returns a vector of pointers to tiles to load. */
+  std::vector<ResultTile*> tiles_to_load() {
+    std::vector<ResultTile*> ret;
+    uint64_t size = 0;
+    for (auto& rt_map : result_tiles_to_load_) {
+      size += rt_map.size();
+    }
+
+    ret.reserve(size);
+    for (auto& rt_map : result_tiles_to_load_) {
+      for (auto& rt : rt_map) {
+        ret.emplace_back(&rt.second);
+      }
+    }
+
+    return ret;
+  }
+
+  /**
+   * Removes duplicate checks for attribute ordering.
+   *
+   * For fragments with adjacent non-empty domains, we only need to check the
+   * interface once. This method marks all fragment bounds as validated except
+   * the bound on the most recent fragment.
+   *
+   * This also marks as validate fragment bounds that are covered or a min/max
+   * for the entire array's non-empty domain, as no check is needed in these
+   * cases.
```

Comment:
This needs to be updated to mention we are finding the fragments that are needed for validation.